### PR TITLE
[wip] Use conda-build 3 for matrix support

### PR DESCRIPTION
**Pull request by msarahan**
@conda-forge/core - first stab here.

As of the first commit here, this is using conda-build's matrix support and outputting folders with trimmed-down individual matrix files in subfolders.  That's how Anaconda's build system works, and I've lifted some code from conda-concourse-ci.

This needs a bit more work, and I'm hoping one of you can help.  Specifically:
- How do I add the new folders that I've generated?  I can certainly hack an add command in at https://github.com/conda-forge/conda-smithy/compare/master...msarahan:conda-build-3?expand=1#diff-d780142cb5b4b67f7a8fd097e056bba6R205, but it seems strange that there isn't something already adding the other files somehow?  I must be missing something.
- This needs to be tested online with each CI to make sure that the template edits I did are actually functional.  Do you have any recommendations on how to set that up?  I suppose I could re-render locally, then do a PR with the re-render, and if that build passes everywhere, it's good?

### CODE REVIEW

**Comment by isuruf** [comment]
> How do I add the new folders that I've generated? I can certainly hack an add command in at https://github.com/conda-forge/conda-smithy/compare/master...msarahan:conda-build-3?expand=1#diff-d780142cb5b4b67f7a8fd097e056bba6R205, but it seems strange that there isn't something already adding the other files somehow? I must be missing something.

Files are added and removed in git automatically with `conda_smithy.feedstocks_io.write_file` and `conda_smithy.feedstocks_io.remove_file`

> This needs to be tested online with each CI to make sure that the template edits I did are actually functional. Do you have any recommendations on how to set that up? I suppose I could re-render locally, then do a PR with the re-render, and if that build passes everywhere, it's good?

Yes, that's what we do. Open a PR in a feedstock that is not `noarch` and one that is `noarch`. If they pass, that's all we need.

**Comment by msarahan** [comment]
> Files are added and removed in git automatically with conda_smithy.feedstocks_io.write_file and conda_smithy.feedstocks_io.remove_file

Thanks, that's exactly what I needed.

**Comment by msarahan** [comment]
Appveyor is looking fine - will need more work on the upload stuff though: https://ci.appveyor.com/project/msarahan/click-feedstock/build/1.0.3/job/6u3mexd9d1wk2325#L655
Actual matrix works fine, though: https://ci.appveyor.com/project/msarahan/click-feedstock/build/1.0.3

Circle: working fine, same as appveyor:
https://circleci.com/gh/msarahan/click-feedstock/9

Travis: *shrug* - still waiting on mac workers...

**Comment by msarahan** [comment]
Travis also ended up coming out fine, aside from the upload stuff: https://travis-ci.org/msarahan/click-feedstock/builds/308655326

**Comment by isuruf** [comment]
@msarahan, thanks for your work on this.
Would it be possible to further strip the packages from the output yaml files that are not related? For example remove `zlib` if the package doesn't use `zlib`?
Reason is that, when the central yaml file is updated for zlib, all feedstocks become outdated and conda-smithy would add these changes in the next rerender. This leads to diffs completely unrelated to the feedstock.

**Comment by msarahan** [comment]
I think so.  I'll look into it.  It might be kind of a pain, though, since fields can be linked.

I really don't think diffs in matrix files are something that anyone should bother looking at anyway.  We already have a culture of people ignoring the CI script changes made in re-rendering, and this is really more of the same.

Anyway, I'll try it and see what happens.

**Comment by jakirkham** [comment]
Also kind of a minor point, but it would be nice if these new directories got a `.` before hand. Would like us to move towards keeping these sorts of things hidden to avoid noise for those using these locally.

**Comment by isuruf** [comment]
@jakirkham, what do you think about moving them to `ci_support` folder?

**Comment by jakirkham** [comment]
Traditionally that has been used for CircleCI stuff. Though maybe we should move the existing content in there to the `.circleci` directory. Had meant to do that before. Otherwise yes that is a good idea.

**Comment by msarahan** [comment]
My latest commit has stuff there for only showing "used" loop variables.  It's currently filtering out anything that doesn't have more than one option.  This is potentially troublesome, because we'd effectively be ignoring all of our pins that have just one entry.

There's two potential fixes:
1. download the central conda_build_config.yaml file (from https://github.com/conda-forge/staged-recipes/pull/4502 or wherever) for every build.  Use this, but override it with the matrix file.  The matrix file will then be a subspace of the central file.
2. expand the matrix file to include anything that gets pinned, even if it has only a single value.

2 is more work, I think, and also means that naming the configurations is more annoying, because you must separate "used" variables into two categories now, and only use the multiple-value ones in the configuration naming (or else face very long configuration names).

Thoughts?

**Comment by isuruf** [comment]
I like 2 because to check whether the feedstock needs to be rebuilt because of an update to a pinned package, you only need to rerender and check if there's a change. Also the maintenance bot only needs to rerender and send a PR if there's a change.
That said, I don't want to burden you too much with more work.

**Comment by jakirkham** [comment]
Think I agree with @isuruf that having full visibility of packages used directly by a dependency.

Couple of questions though. First would this include transitive dependencies? For example, if I depend on `python`, which depends on `zlib`, would `zlib` get included if I didn't explicitly list it as a dependency or would it get left out as I didn't list it? Second what challenges do you see with second route? Is it only the naming issue or are there other challenges we might need to address.

**Comment by msarahan** [comment]
No transitive deps.  Only direct deps.

I have something working, but it'll take another conda-build release to make it available.

**Comment by jakirkham** [comment]
Regarding transitive dependencies, great!

Ah so is the naming something `conda-build` will address then?

Something we have been unable to do with the `conda-build-all`-based `conda-smithy` is warn and/or error if a configuration conflicts in such a way as to disable the build totally. Is that something we can address with `conda-build` 3.0?

**Comment by isuruf** [comment]
I'm not sure I understand about transitive deps. Why do we want to worry about zlib for python packages?

**Comment by msarahan** [comment]
You don't want to worry about zlib for python packages.  That's why we ignore transitive deps.

You can have conda-build 3 "finalize" recipes when it's rendering them.  It slows things down a lot, but perhaps it's still a net win for preventing dud builds.  This can change by tweaking the flags at https://github.com/conda-forge/conda-smithy/pull/618/files#diff-3a2c4a048580db27ad9db93e2780b511R181

If builds are unsatisfiable, it'll error out.  Usually this is because things are not available, but if you have conflicting pins, it should also error out in the same way.

**Comment by jakirkham** [comment]
As an example, `click` should not care about `zlib`. If `click` started using `zlib` and depended on the version, we would want to add it as an explicit dependency and then it would be tracked.

Thanks for the pointer. Seems easy enough.

Though if `conda-build` is already erroring out on unsatisfiable dependencies, I think that is probably sufficient for solving that issue.

**Comment by msarahan** [comment]
As another example, consider libtiff:

https://github.com/anacondarecipes/libtiff-feedstock/blob/master/recipe/meta.yaml

there's more than one compiler on windows, so that's a loop variable.  However, jpeg and zlib have only one value.  Those values needs to be recorded, along with the compiler value, but only the compiler value should be part of the configuration name.

The configuration:
```
c_compiler:
- vs2008
cxx_compiler:
- vs2008
jpeg:
- '9'
pin_run_as_build:
  libboost:
    max_pin: x.x
  python:
    max_pin: x.x
    min_pin: x.x
  r-base:
    max_pin: x.x.x
    min_pin: x.x.x
target_platform:
- win-64
zip_keys:
- - pyqt
  - qt
- - python
  - ipython
  - vc
  - c_compiler
  - cxx_compiler
  - fortran_compiler_version
zlib:
- '1.2'
```

The configuration name: ```c_compilervs2008cxx_compilervs2008```

File: conda_smithy/templates/appveyor.yml.tmpl [hunk 1]
```diff
@@ -47,12 +47,15 @@ install:
     {% if build_setup -%}
     {{ build_setup }}{% endif %}
 
+    # testing purposes: get conda-build from defaults
+    - cmd: conda install -n root --quiet --yes -c defaults conda-build
```
Comment (msarahan) [inline conda_smithy/templates/appveyor.yml.tmpl]: Remove this line before merging (version constraint should come from conda-forge-build-setup)

File: conda_smithy/templates/run_docker_build.tmpl [hunk 2]
```diff
@@ -72,9 +59,12 @@ conda install --yes --quiet conda-forge-build-setup
 {% if build_setup -%}
 {{ build_setup }}{% endif -%}
 
-conda build /recipe_root --quiet || exit 1
+# testing purposes: get conda-build from defaults
+conda install -n root --quiet --yes -c defaults conda-build
```
Comment (msarahan) [inline conda_smithy/templates/run_docker_build.tmpl]: Remove this line before merging (version constraint should come from conda-forge-build-setup)

File: conda_smithy/templates/travis.yml.tmpl [hunk 3]
```diff
@@ -68,9 +63,10 @@ install:
       conda install --yes --quiet conda-forge-build-setup
       {% if build_setup -%}
       {{ build_setup }}{% endif %}
+      conda install -n root --quiet --yes -c defaults conda-build
```
Comment (msarahan) [inline conda_smithy/templates/travis.yml.tmpl]: Remove this line before merging (version constraint should come from conda-forge-build-setup)

**Comment by msarahan** [comment]
@conda-forge/core this is ready for review.  This PR requires conda-build 3.1.1 (available on defaults).  This PR is self-contained, but I'm expecting that https://github.com/conda-forge/staged-recipes/pull/4502 or something like it will be used to provide the conda_build_config.yaml file at rendering time.

**Comment by msarahan** [comment]
I'm going to close this, in case it's noisy.  I'll reopen it and rebase when I have py3 coverage working.

**Comment by jakirkham** [comment]
No worries. It is not noisy. Please feel free to keep it open if it makes it easier to push forward.

**Pull request closed without merging**